### ISSUE TITLE
Pylance spends too much time when using Pony ORM

### ISSUE THREAD

**Issue by onion7878** [microsoft/pylance-release#1049]
## Environment data

-   Language Server version: Pylance language server 2021.3.1 (pyright 5d6db8d1) starting
-   OS and version: Microsoft Windows [Version 10.0.19042.867]
-   Python version: Python 3.8.6 (use poetry manage virtual environment) 

## Expected behaviour

Pylance analyses the file quickly

## Actual behaviour

Pylance took close to 10 minutes to analyze, not sure if pylancer or Pony ORM is the cause

## Logs

```
[Info  - 8:09:03 PM] Pylance language server 2021.3.1 (pyright 5d6db8d1) starting
[Info  - 8:09:04 PM] Server root directory: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist
[Info  - 8:09:04 PM] No configuration file found.
[Info  - 8:09:04 PM] Setting pythonPath for service "test_pylance": "c:\Users\Hanlin\Desktop\test_pylance\.venv\Scripts\python.exe"
Search paths found for configured python interpreter:
  C:\Program Files\Python38\DLLs
  C:\Program Files\Python38\lib
  C:\Program Files\Python38
  c:\Users\Hanlin\Desktop\test_pylance\.venv
  c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages
[Error - 8:09:04 PM] stubPath c:\Users\Hanlin\Desktop\test_pylance\typings is not a valid directory.
[Info  - 8:09:04 PM] Assuming Python version 3.8
[Info  - 8:09:04 PM] Assuming Python platform Windows
[Info  - 8:09:04 PM] Searching for source files
[Info  - 8:09:04 PM] Auto-excluding c:\Users\Hanlin\Desktop\test_pylance\.venv
[Info  - 8:09:04 PM] Found 1 source file
[FG] parsing: c:\Users\Hanlin\Desktop\test_pylance\demo.py (19ms)
[FG] parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 2ms] (105ms)
[FG] binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\builtins.pyi (41ms)
[FG] binding: c:\Users\Hanlin\Desktop\test_pylance\demo.py ...
[FG]   parsing: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\orm\__init__.py [fs read 0ms] (6ms)
[FG]   binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\orm\__init__.py ...
[FG]     parsing: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\orm\core.py [fs read 3ms] (212ms)
[FG]     binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\orm\core.py (92ms)
[FG]   binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\orm\__init__.py (306ms)
[FG] binding: c:\Users\Hanlin\Desktop\test_pylance\demo.py (313ms)
[Info  - 8:09:04 PM] Background analysis(1) root directory: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist
[Info  - 8:09:04 PM] Background analysis(1) started
Background analysis message: setConfigOptions
Background analysis message: setTrackedFiles
Background analysis message: markAllFilesDirty
Background analysis message: ensurePartialStubPackages
Background analysis message: setFileOpened
Background analysis message: getSemanticTokens
[BG(1)] parsing: c:\Users\Hanlin\Desktop\test_pylance\demo.py (15ms)
[BG(1)] parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 3ms] (81ms)
[BG(1)] binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\builtins.pyi (32ms)
[BG(1)] binding: c:\Users\Hanlin\Desktop\test_pylance\demo.py ...
[BG(1)]   parsing: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\orm\__init__.py [fs read 0ms] (3ms)
IntelliCode model c:\Users\Hanlin\.vscode\extensions\visualstudioexptteam.vscodeintellicode-1.2.11\cache\E61945A9A512ED5E1A3EE3F1A2365B88F8FE_E4E9EADA96734F01970E616FAB2FAC19
Loading ONNX runtime...
Loaded ONNX runtime. Creating IntelliCode session...
2021-03-15 20:09:04.8634478 [I:onnxruntime:, inference_session.cc:174 onnxruntime::InferenceSession::ConstructorCommon] Creating and using per session threadpools since use_per_session_threads_ is true
2021-03-15 20:09:04.8649136 [I:onnxruntime:, inference_session.cc:840 onnxruntime::InferenceSession::Initialize] Initializing session.
2021-03-15 20:09:04.8650039 [I:onnxruntime:, inference_session.cc:865 onnxruntime::InferenceSession::Initialize] Adding default CPU execution provider.
2021-03-15 20:09:04.8660784 [I:onnxruntime:, reshape_fusion.cc:37 onnxruntime::ReshapeFusion::ApplyImpl] Total fused reshape node count: 0
2021-03-15 20:09:04.8665587 [I:onnxruntime:, reshape_fusion.cc:37 onnxruntime::ReshapeFusion::ApplyImpl] Total fused reshape node count: 0
2021-03-15 20:09:04.8679674 [V:onnxruntime:, inference_session.cc:679 onnxruntime::InferenceSession::TransformGraph] Node placements
2021-03-15 20:09:04.8680356 [V:onnxruntime:, inference_session.cc:681 onnxruntime::InferenceSession::TransformGraph] All nodes have been placed on [CPUExecutionProvider].
2021-03-15 20:09:04.8681080 [V:onnxruntime:, session_state.cc:71 onnxruntime::SessionState::CreateGraphInfo] SaveMLValueNameIndexMapping
2021-03-15 20:09:04.8682055 [V:onnxruntime:, session_state.cc:116 onnxruntime::SessionState::CreateGraphInfo] Done saving OrtValue mappings.
2021-03-15 20:09:04.8690477 [I:onnxruntime:, finalize_session_state.cc:173 onnxruntime::SaveInitializedTensors] Saving initialized tensors.
2021-03-15 20:09:04.8779093 [I:onnxruntime:, finalize_session_state.cc:225 onnxruntime::SaveInitializedTensors] Done saving initialized tensors
2021-03-15 20:09:04.8812363 [I:onnxruntime:, inference_session.cc:954 onnxruntime::InferenceSession::Initialize] Session successfully initialized.
Created IntelliCode session.
[BG(1)]   binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\orm\__init__.py ...
[BG(1)]     parsing: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\orm\core.py [fs read 2ms] (230ms)
[BG(1)]     binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\orm\core.py (84ms)
[BG(1)]   binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\orm\__init__.py (316ms)
[BG(1)] binding: c:\Users\Hanlin\Desktop\test_pylance\demo.py (321ms)
[BG(1)] parsing: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\__init__.py [fs read 1ms] (17ms)
[BG(1)] binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\__init__.py (1ms)
[BG(1)] parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi [fs read 0ms] (2ms)
[BG(1)] binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi (0ms)
[BG(1)] parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\typing.pyi [fs read 0ms] (18ms)
[BG(1)] binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\typing.pyi (6ms)
[BG(1)] parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stubs\typing-extensions\typing_extensions.pyi [fs read 1ms] (2ms)
[BG(1)] binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stubs\typing-extensions\typing_extensions.pyi (0ms)
[BG(1)] parsing: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\utils\__init__.py [fs read 1ms] (1ms)
[BG(1)] binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\utils\__init__.py ...
[BG(1)]   parsing: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\utils\utils.py [fs read 0ms] (9ms)
[BG(1)]   binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\utils\utils.py (4ms)
[BG(1)]   parsing: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\utils\properties.py [fs read 0ms] (1ms)
[BG(1)]   binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\utils\properties.py (1ms)
[BG(1)] binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\utils\__init__.py (15ms)
[BG(1)] parsing: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\thirdparty\decorator.py [fs read 0ms] (6ms)
[BG(1)] binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\thirdparty\decorator.py (2ms)
[BG(1)] parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\inspect.pyi [fs read 0ms] (4ms)
[BG(1)] binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\inspect.pyi (2ms)
[BG(1)] parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\functools.pyi [fs read 0ms] (4ms)
[BG(1)] binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\functools.pyi (1ms)
[BG(1)] parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\abc.pyi [fs read 1ms] (1ms)
[BG(1)] binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\abc.pyi (1ms)
[BG(1)] parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\sys.pyi [fs read 1ms] (3ms)
[BG(1)] binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\sys.pyi (6ms)
[BG(1)] parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\types.pyi [fs read 0ms] (4ms)
[BG(1)] binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\types.pyi (2ms)
[BG(1)] parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\ast.pyi [fs read 0ms] (4ms)
[BG(1)] binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\ast.pyi ...
[BG(1)]   parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\_ast.pyi [fs read 0ms] (3ms)
[BG(1)]   binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\_ast.pyi (3ms)
[BG(1)] binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\ast.pyi (9ms)
[BG(1)] parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\re.pyi [fs read 0ms] (1ms)
[BG(1)] binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\re.pyi (3ms)
[BG(1)] parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\enum.pyi [fs read 0ms] (1ms)
[BG(1)] binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\enum.pyi (0ms)
[BG(1)] parsing: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\py23compat.py [fs read 0ms] (4ms)
[BG(1)] binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\py23compat.py (0ms)
[BG(1)] parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\itertools.pyi [fs read 0ms] (3ms)
[BG(1)] binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\itertools.pyi (1ms)
[BG(1)] parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\threading.pyi [fs read 0ms] (1ms)
[BG(1)] binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\threading.pyi (2ms)
[BG(1)] parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\logging\__init__.pyi [fs read 0ms] (6ms)
[BG(1)] binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\logging\__init__.pyi (7ms)
[BG(1)] parsing: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\orm\decompiling.py [fs read 1ms] (12ms)
[BG(1)] binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\orm\decompiling.py (7ms)
[BG(1)] parsing: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\thirdparty\compiler\__init__.py [fs read 0ms] (1ms)
[BG(1)] binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\thirdparty\compiler\__init__.py (0ms)
[BG(1)] parsing: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\thirdparty\compiler\transformer.py [fs read 1ms] (39ms)
[BG(1)] binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\thirdparty\compiler\transformer.py ...
[BG(1)]   parsing: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\thirdparty\compiler\ast.py [fs read 1ms] (22ms)
[BG(1)]   binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\thirdparty\compiler\ast.py (10ms)
[BG(1)] binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\thirdparty\compiler\transformer.py (46ms)
Background analysis message: getDiagnosticsForRange
Background analysis message: analyze
[BG(1)] analyzing: c:\Users\Hanlin\Desktop\test_pylance\demo.py ...
[BG(1)]   checking: c:\Users\Hanlin\Desktop\test_pylance\demo.py ...
[BG(1)]     parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\pickle.pyi [fs read 0ms] (3ms)
[BG(1)]     binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\pickle.pyi (2ms)
[BG(1)]     parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\io.pyi [fs read 0ms] (5ms)
[BG(1)]     binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\io.pyi (1ms)
[BG(1)]     parsing: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\orm\ormtypes.py [fs read 0ms] (11ms)
[BG(1)]     binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\orm\ormtypes.py (3ms)
[BG(1)]     parsing: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\orm\asttranslation.py [fs read 0ms] (3ms)
[BG(1)]     binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\orm\asttranslation.py (7ms)
[BG(1)]     parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\platform.pyi [fs read 0ms] (2ms)
[BG(1)]     binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\platform.pyi (0ms)
[BG(1)]     parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\collections\__init__.pyi [fs read 1ms] (4ms)
[BG(1)]     binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\collections\__init__.pyi ...
[BG(1)]       parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\_collections_abc.pyi [fs read 0ms] (0ms)
[BG(1)]       binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\_collections_abc.pyi (0ms)
[BG(1)]     binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\collections\__init__.pyi (2ms)
[BG(1)]     parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\opcode.pyi [fs read 0ms] (1ms)
[BG(1)]     binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\opcode.pyi (0ms)
[BG(1)]     parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\symbol.pyi [fs read 0ms] (3ms)
[BG(1)]     binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\symbol.pyi (1ms)
[BG(1)]     parsing: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\parser.pyi [fs read 0ms] (1ms)
[BG(1)]     binding: c:\Users\Hanlin\.vscode\extensions\ms-python.vscode-pylance-2021.3.1\dist\typeshed-fallback\stdlib\parser.pyi (0ms)
[BG(1)]     parsing: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\thirdparty\compiler\consts.py [fs read 0ms] (0ms)
[BG(1)]     binding: c:\Users\Hanlin\Desktop\test_pylance\.venv\lib\site-packages\pony\thirdparty\compiler\consts.py (1ms)
[BG(1)]   checking: c:\Users\Hanlin\Desktop\test_pylance\demo.py (562405ms)
[Info  - 8:18:27 PM] [BG(1)] Long operation: checking: c:\Users\Hanlin\Desktop\test_pylance\demo.py (562405ms)
[BG(1)] analyzing: c:\Users\Hanlin\Desktop\test_pylance\demo.py (562405ms)
[Info  - 8:18:27 PM] [BG(1)] Long operation: analyzing: c:\Users\Hanlin\Desktop\test_pylance\demo.py (562405ms)
```

## Code Snippet / Additional information

demo.py

```python
# from pony import orm
from pony.orm import *

if __name__ == "__main__":
    q = select()
```

I have tested that this issue is triggered when using the select() function

**Comment by onion7878** [comment]
and Pony ORM version is 0.7.14

```bash
$ poetry show pony
name         : pony
version      : 0.7.14
description  : Pony Object-Relational Mapper
```

**Comment by erictraut** [comment]
Thanks for the bug report. A similar problem was reported recently, and I added some additional internal caching to avoid duplicate type evaluation. With this additional caching, the analysis time goes from 562000ms down to 176ms! This fix will be in the next release of pylance.

**Comment by onion7878** [comment]
ok, wait for next release.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.3.2, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202132-17-march-2021

**Comment by onion7878** [comment]
I just tested it and it has solved the problem, thx 

```log
[BG(1)]   checking: c:\Users\Hanlin\Desktop\test_pylance\demo.py (222ms)
[BG(1)] analyzing: c:\Users\Hanlin\Desktop\test_pylance\demo.py (224ms)
```
